sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala

```diff
@@ -155,8 +156,9 @@ private[sql] class CacheManager extends Logging {
    * function will over invalidate.
    */
   private[sql] def invalidateCache(plan: LogicalPlan): Unit = writeLock {
+    val canonicalized = plan.canonicalized
     cachedData.foreach {
-      case data if data.plan.collect { case p if p.sameResult(plan) => p }.nonEmpty =>
+      case data if data.plan.collect { case p if p.sameResult(canonicalized) => p }.nonEmpty =>
```

Comment:
I don't think so.
For example, if the cached plan is `LocalRelation` (which is canonicalized) and the `plan` argument is `SubqueryAlias(LocalRelation)` (which is not canonicalized), it will fail to find the same-result plan.